go-controller/pkg/node/node.go

```diff
@@ -292,33 +292,52 @@ func (n *OvnNode) WatchEndpoints() error {
 			if reflect.DeepEqual(epNew.Subsets, epOld.Subsets) {
 				return
 			}
-			newEndpointAddressMap := buildEndpointAddressMap(epNew.Subsets)
-			for _, subset := range epOld.Subsets {
-				for _, address := range subset.Addresses {
-					if _, ok := newEndpointAddressMap[address.IP]; !ok {
-						deleteConntrack(address.IP)
+			newEpAddressMap := buildEndpointAddressMap(epNew.Subsets)
+			for item := range buildEndpointAddressMap(epOld.Subsets) {
+				if _, ok := newEpAddressMap[item]; !ok {
+					err := deleteConntrack(item.ip, item.port, item.protocol)
+					if err != nil {
+						klog.Errorf("Failed to delete conntrack entry for %s: %v", item.ip, err)
 					}
 				}
 			}
 		},
 		DeleteFunc: func(obj interface{}) {
 			ep := obj.(*kapi.Endpoints)
-			for _, subset := range ep.Subsets {
-				for _, address := range subset.Addresses {
-					deleteConntrack(address.IP)
+			for item := range buildEndpointAddressMap(ep.Subsets) {
+				err := deleteConntrack(item.ip, item.port, item.protocol)
+				if err != nil {
+					klog.Errorf("Failed to delete conntrack entry for %s: %v", item.ip, err)
 				}
+
 			}
 		},
 	}, nil)
 	return err
 }
 
-func buildEndpointAddressMap(epSubsets []kapi.EndpointSubset) map[string]struct{} {
-	addressMap := make(map[string]struct{})
+type epAddressItem struct {
+	ip       string
+	port     int32
+	protocol kapi.Protocol
+}
+
+//buildEndpointAddressMap builds a map of all UDP and SCTP ports in the endpoint subset along with that port's IP address
+func buildEndpointAddressMap(epSubsets []kapi.EndpointSubset) map[*epAddressItem]struct{} {
+	epMap := make(map[*epAddressItem]struct{})
```

Comment:
@JacobTanenbaum I led you a bit astray... Turns out we need just `epAddressItem` here (eg not a pointer to one) becuase otherwise the map access above for Update/Delete won't return true for the same value (since it's comparing the pointer, not the object's value).

```
package main

import (
	"fmt"
)

type foo struct {
	ip string
}

func main() {
	a := make(map[*foo]struct{})
	a[&foo{ip:"1.1.1.1"}] = struct{}{}
	_, ok := a[&foo{ip:"1.1.1.1"}]
	fmt.Printf("* found %v\n", ok)

	b := make(map[foo]struct{})
	b[foo{ip:"1.1.1.1"}] = struct{}{}
	_, ok = b[foo{ip:"1.1.1.1"}]
	fmt.Printf("not-* found %v\n", ok)
}
```
gives you:
```
* found false
not-* found true
```

Reply:
done